memcache/memcache.go

```diff
@@ -711,3 +747,238 @@ func (c *Client) incrDecr(verb, key string, delta uint64) (uint64, error) {
+approach to setting values in memcached- instead of having multiple methods each do different things, you can consolidate
+everything you wish to do in one line with the meta command and its array of flags.
+
+Arguments
+@metaItem: encapsulates the key of the item, its new value, and all flags to apply to the meta set command
+
+Return values
+@metaDataResponse: encapsulates all values returned as a result of the flags which return a value applied to the meta set command
+@err: error that can be raised as a result of the operation- errors include: error because value wasn't stored, error because of cache miss,
+error because the item alerady exists (occurs with certain flags), I/O errors, malformed key error, and generic error for an unknown response
+from memcached
+*/
+func (c *Client) MetaSet(metaItem *MetaSetItem) (metaDataResponse *MetaResponseMetadata, err error) {
+	err = c.onMetaItem(metaItem, (*Client).processMetaSet, func(metaData *MetaResponseMetadata) { metaDataResponse = metaData })
+	return
+}
+
+func (c *Client) onMetaItem(item *MetaSetItem, fn func(*Client, *bufio.ReadWriter, *MetaSetItem, func(metaData *MetaResponseMetadata)) error, cb func(metaData *MetaResponseMetadata)) error {
+	addr, err := c.selector.PickServer(item.Key)
+	if err != nil {
+		return err
+	}
+	cn, err := c.getConn(addr)
+	if err != nil {
+		return err
+	}
+	defer cn.condRelease(&err)
+	if err = fn(c, cn.rw, item, cb); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) processMetaSet(rw *bufio.ReadWriter, item *MetaSetItem, cb func(*MetaResponseMetadata)) error {
+	if !legalKey(item.Key) {
+		return ErrMalformedKey
+	}
+
+	command := c.parseFlagsForMetaSet(item)
+
+	var err error
+	_, err = fmt.Fprintf(rw, command)
+
+	if err != nil {
+		return nil
+	}
+	if _, err = rw.Write(item.Value); err != nil {
+		return err
+	}
+	if _, err := rw.Write(crlf); err != nil {
+		return err
+	}
+	if err := rw.Flush(); err != nil {
+		return err
+	}
+	response, err := rw.ReadSlice('\n')
+	if err != nil {
+		return err
+	}
+	switch {
+	case bytes.HasPrefix(response, metaResultStored):
+		// the first two characters are being processed in this switch-case block, the other cases are all errors
+		// so we don't need to save them into the meta data object
+		responseMetadataComponents := strings.Fields(string(response))[1:]
+		metaResponseMetadata := new(MetaResponseMetadata)
+		if err = parseMetaResponseMetadata(responseMetadataComponents, metaResponseMetadata); err != nil {
+			return err
+		}
+		cb(metaResponseMetadata)
+		return nil
+	case bytes.HasPrefix(response, metaResultNotStored):
+		return ErrNotStored
+	case bytes.HasPrefix(response, metaResultExists):
+		return ErrCASConflict
+	case bytes.HasPrefix(response, metaResultNotFound):
+		return ErrCacheMiss
+	}
+	return fmt.Errorf("memcache: unexpected response line from ms: %q", string(response))
```

Comment:
Minor: Probably we can use switch default here ?

Reply:
I don't think it's necessary, if none of the cases are caught, then it will get to this line no matter what.